### ISSUE TITLE
effective_limits::ErrorKind exposes sys-info implementation details

### ISSUE THREAD

**Issue by jclulow** [rbtcollins/effective-limits.rs#3]
The [`ErrorKind`](https://docs.rs/effective-limits/0.3.0/effective_limits/enum.ErrorKind.html) enumeration includes a `SysInfo` variant, used when we cannot obtain the total memory via the **sys-info** crate.  The **sys-info** crate is quite large, but does not support every platform that Rust supports.  For other platforms, we could use conditional compilation to swap that crate out for another one -- except that including the `sys_info::Error` type in the `ErrorKind` enum prevents us from doing that.

Would you be OK with a small breaking change to the API, replacing the current `SysInfo` variant with a string-based variant for memory size detection failures?  That way, on platforms where **sys-info** is not compiled in, we can still render an error about the failure to find the size of total memory without needing to use that crate or conditionally change the API.

**Comment by rbtcollins** [comment]
Hey, so I think switching to anyhow and this_error is probably a good idea anyway, even though its not the crux of this ticket. I'm very happy to accommodate making this crate accessible to other rust platforms - though I will note that as we use it in rustup, we know that its available to all tier 1 platforms right now, and if the platform you are working on wants to become tier 1 we'll need to solve the availability of sys_info for it at some point anyway.

In terms of how to do it, the reason I mention this_error is that I would prefer not to discard the source details on platforms where sys_info is available, since that makes diagnostics harder for users; instead, conditionally define the API, so that only affected platforms are affected.

This would look something like:
- a feature sys_info in cargo.toml
- default-features ['sys_info']
- 

```
    #[cfg(sys_info)]
    #error[("sys-info failure")]
     MemInfoLookupFailure{
         #[source]
         source: SysInfo::Error
    },
    #[cfg(not(sys_info))]
    #error[("alternative-thing failure")]
     MemInfoLookupFailure{
         #[source]
         source: Alternative::Error
    },
  ```

then callers that want to catch MemInfoLookupFailure should be able to; the underlying error they'll have to make multiple branches for if they care about non-tier-1 platforms, but generic print-and-move-on code will work fine.

**Comment by jclulow** [comment]
I'm actually working on getting a presently tier 2 platform (illumos/Solaris) to have full binary availability in `rustup` at the moment!  I've managed to knock together a complete prototype that can download and install a nightly toolchain on an illumos system, and I'm currently working on various pull requests to get it all upstream.

I have added illumos support to **sys-info** in an open PR: https://github.com/FillZpp/sys-info-rs/pull/51

I'm conscious of the fact that it's a bigger change that will require longer to review, which is why I started looking into this angle as well.  I'd love for us to eventually be Tier 1, so it seems like I should keep pushing on the **sys-info** PR for a bit longer for now.

Thanks for taking a look!
